File: josephus.py
def is_power_of_two(n):
    return (n != 0) and (n & (n-1) == 0)

def highest_poweroftwo(n):
    result = 0
    
    for i in range(n, 0, -1):
        if is_power_of_two(i):
            result = i
            break
    return result
    
def josephus(n):
    if n == 0:
        print("No seats, winning seat cant be found.")
    if n < 0:
        print("You cannot have a negative number of seats")
    if is_power_of_two(n):
        print("Winning seat is 1.")
    else:    
    # where n = 2a(highest power of 2 less than n) - l    
        l = n - highest_poweroftwo(n)
    # winning seat is 2(l) + 1
        l = 2 * l  + 1
        print("The winning seat number is: ", l)
    
    
josephus(90) 

File: josephus.c
#include<stdio.h> 
#include<stdbool.h> 
#include<math.h> 

bool is_power_of_two(int n){
    if (n ==0)
        return false;
    while (n != 1){
        if (n % 2 == 0)
            return false;
        n = n/2;
    }
    return true;
}

int highest_poweroftwo(int n){
    int result = 0;

    for (int i = n; i >= 1; i--){
        if ((i & (i - 1)) == 0){
            result = i;
            break;
        }
    }
    return result;
}

int josephus(int n){

    if (n == 0){
        printf("No seats, winning seat cant be found.\n");
        return 0;
    }
    if (n < 0){
        printf("You cannot have a negative number of seats.\n");
        return 0;
    }
    if (is_power_of_two(n))
        return 1;
    else
    {
        int l;
        l = n - highest_poweroftwo(n);
        l = 2 * l + 1;
        return (l);
    }
}

int main(){
    int ans;
    ans = josephus(90); //change the number you want to solve for here
    printf("%s %i\n", "The winning seat is\n", ans);

    return 0;
}